### PR TITLE
RS/YJ/Rule 11-14

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
I've tested this rule with Juan's testing JSON.

### CODE REVIEW

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+            def manual_check_required(self, context, calc_vals=None, data=None):
+                piping_info_b = data["piping_info_b"]
+
+                return any(not piping["IS_RECIRC"] for piping in piping_info_b.values())
+
+            def get_calc_vals(self, context, data=None):
+                piping_info_b = data["piping_info_b"]
+
+                return {"piping_info_id_b": piping_info_b}
```
Comment (yunjoonjung-PNNL) [inline rct229/rulesets/ashrae9012019/section11/section11rule14.py:127]: The type of ```piping_info_id_b``` is dict[str, dict]. What would be the best way to pass this dict over to the ```rule_check``` considering the reporting?

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                        for pump_id in swh_comps_dict_b[swh_dist_sys_b["id"]].pumps:
+                            pump = find_one_with_field_value(
+                                "$.pumps[*]",
+                                "id",
+                                pump_id,
+                                rmd_b,
+                            )
+                            piping_info_b[piping_id]["PUMP_POWER"] = pump.get(
+                                "design_electric_power"
+                            )
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section11/section11rule14.py:104]: should this be `+=`?

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                piping_info_b = data["piping_info_b"]
+
+                return all(
+                    piping["IS_RECIRC"] for piping in piping_info_b.values()
+                ) and all(
+                    piping["PUMP_POWER"] > MIN_PUMP_POWER
+                    for piping in piping_info_b.values()
+                )
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section11/section11rule14.py:137]: can do return 
```python
all([ piping["IS_RECIRC"] and piping["PUMP_POWER"] > MIN_PUMP_POWER for piping in piping_info_b.values()])
```

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                super(Section11Rule14.RMDRule.SWHDistributionRule, self).__init__(
+                    rmds_used=produce_ruleset_model_description(
+                        USER=False, BASELINE_0=True, PROPOSED=True
+                    ),
+                )
+
+            def manual_check_required(self, context, calc_vals=None, data=None):
+                piping_info_b = data["piping_info_b"]
+
+                return any(not piping["IS_RECIRC"] for piping in piping_info_b.values())
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section11/section11rule14.py:122]: I think this is `all`?
Using `any` would have possibility to return True if only one of the pipings meet the failed condition `piping.get("IS_RECIRC") and piping.get("PUMP_POWER", 0) == 0 for entry in hash.values() = False`

I think the correct one is `all(not piping["IS_RECIRC"] or piping["PUMP_POWER"] > MIN_PUMP_POWER for piping in piping_info_b.values()) and not all([ piping["IS_RECIRC"] and piping["PUMP_POWER"] > MIN_PUMP_POWER for piping in piping_info_b.values()])`
I am not sure how to simply this yet. may need some more effort.

**Review by weilixu — CHANGES_REQUESTED** [review]

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py (hunk 3 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:137]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py (hunk 2 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:104]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                    )
+                    for swh_dist_sys_id_b in swh_comps_dict_b[swh_dist_id_b].uses
+                    for swh_use_p in find_all(
+                        f'$.buildings[*].building_segments[*].zones[*].spaces[*].service_water_heating_uses[*][?(@.id="{swh_dist_sys_id_b}")].use',
+                        rmd_p,
+                    )
+                ]
+            )
+
+            return swh_use_loads_p
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:68]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+            swh_use_loads_p = all(
+                [
+                    True if swh_use_p > 0.0 else False
+                    for swh_dist_id_b in find_all(
+                        "$.service_water_heating_distribution_systems[*].id", rmd_b
+                    )
+                    for swh_dist_sys_id_b in swh_comps_dict_b[swh_dist_id_b].uses
+                    for swh_use_p in find_all(
+                        f'$.buildings[*].building_segments[*].zones[*].spaces[*].service_water_heating_uses[*][?(@.id="{swh_dist_sys_id_b}")].use',
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:62]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+                return any(not piping["IS_RECIRC"] for piping in piping_info_b.values())
+
+            def get_calc_vals(self, context, data=None):
+                piping_info_b = data["piping_info_b"]
+
+                return {"piping_info_id_b": piping_info_b}
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                piping_info_b = data["piping_info_b"]
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:130]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py (hunk 1 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:127]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:62]: Good point. Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py (hunk 4 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py:126]: Good point! Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule14.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            for swh_dist_sys_b in find_all(
+                "$.service_water_heating_distribution_systems[*]", rmd_b
+            ):
+                for piping_id_b in swh_comps_dict_b[swh_dist_sys_b["id"]].piping:
+                    piping = find_one_with_field_value(
+                        "$.service_water_heating_distribution_systems[*].service_water_piping[*]",
+                        "id",
+                        piping_id_b,
+                        rmd_b,
+                    )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule14.py]: Good point. Addressed.

**Review by weilixu — APPROVED** [review]

**Pull request merged**